File: jugglerhelper/src/main/java/net/maxsmr/jugglerhelper/activities/BaseJugglerActivity.java
package net.maxsmr.jugglerhelper.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MotionEvent;

import androidx.annotation.AnimRes;
import androidx.core.util.Pair;
import androidx.fragment.app.Fragment;

import net.maxsmr.commonutils.android.gui.FragmentFinder;
import net.maxsmr.jugglerhelper.fragments.BaseJugglerFragment;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.List;

import me.ilich.juggler.Navigable;
import me.ilich.juggler.change.Add;
import me.ilich.juggler.gui.JugglerActivity;
import me.ilich.juggler.states.State;

import static me.ilich.juggler.Juggler.DATA_ANIMATION_FINISH_ENTER;
import static me.ilich.juggler.Juggler.DATA_ANIMATION_FINISH_EXIT;

public class BaseJugglerActivity extends JugglerActivity {

    @Nullable
    protected State<?> initialState;

    @Nullable
    protected Bundle savedInstanceState;

    private boolean isCommitAllowed = false;

    private boolean isResumed = false;

    public boolean isCommitAllowed() {
        return isCommitAllowed;
    }

    public boolean isActivityResumed() {
        return isResumed;
    }

    public boolean isActivityDestroyed() {
        return (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 && !isDestroyed());
    }

    @AnimRes
    protected int getFinishEnterAnimataion() {
        return 0;
    }

    @AnimRes
    protected int getFinishExitAnimataion() {
        return 0;
    }

    protected boolean shouldActivateState(State<?> state) {
        return true;
    }

    protected void doActionStateNotActivated(State<?> state) {

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        getIntent().putExtra(DATA_ANIMATION_FINISH_ENTER, getFinishEnterAnimataion());
        getIntent().putExtra(DATA_ANIMATION_FINISH_EXIT, getFinishExitAnimataion());
        initialState = getActivityState();
        if (!shouldActivateState(initialState)) {
            getIntent().putExtra(EXTRA_STATE, (Serializable) null);
            doActionStateNotActivated(initialState);
        }
        this.savedInstanceState = savedInstanceState;
        isCommitAllowed = true;
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        savedInstanceState = outState;
        isCommitAllowed = false;
    }

    @Override
    protected void onResumeFragments() {
        super.onResumeFragments();
        isCommitAllowed = true;
    }

    @Override
    protected void onPause() {
        super.onPause();
        isResumed = false;
    }

    @Override
    protected void onResume() {
        super.onResume();
        isResumed = true;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
//        logger.debug("onTouchEvent(), event=" + event);
        List<Fragment> fragments = getSupportFragmentManager().getFragments();
        for (Fragment f : fragments) {
            if (f instanceof BaseJugglerFragment && f.isAdded()) {
                ((BaseJugglerFragment) f).onTouchEvent(event);
            }
        }
        return super.onTouchEvent(event);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
//        logger.debug("onKeyDown(), keyCode=" + keyCode + ", event=" + event);
        List<Fragment> fragments = getSupportFragmentManager().getFragments();
        for (Fragment f : fragments) {
            if (f instanceof BaseJugglerFragment && f.isAdded()) {
                ((BaseJugglerFragment) f).onKeyDown(keyCode, event);
            }
        }
        return super.onKeyDown(keyCode, event);
    }

    @Nullable
    public Fragment findFragmentById(int id) {
        Pair<Integer, Fragment> fragment = FragmentFinder.findFragmentById(getSupportFragmentManager(), id);
        if (fragment != null) {
            return fragment.second;
        }
        return null;
    }

    @Nullable
    public Fragment findFragmentByTag(String tag) {
        Pair<Integer, Fragment> fragment = FragmentFinder.findFragmentByTag(getSupportFragmentManager(), tag);
        if (fragment != null) {
            return fragment.second;
        }
        return null;
    }

    @Nullable
    public <F extends Fragment> F findFragmentByClass(Class<F> clazz) {
        Pair<Integer, F> fragment = FragmentFinder.findFragmentByClass(getSupportFragmentManager(), clazz);
        if (fragment != null) {
            return fragment.second;
        }
        return null;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        List<Fragment> fragments = getSupportFragmentManager().getFragments();
        for (Fragment f : fragments) {
            if (f != null && f.isAdded()) {
                f.onRequestPermissionsResult(requestCode, permissions, grantResults);
            }
        }
    }

    @Nullable
    public State getActivityState() {
        final Intent intent = getIntent();
        if (intent == null) {
            throw new IllegalStateException(Intent.class.getSimpleName() + " is null");
        }
        final State<?> state;
        if (intent.hasExtra(EXTRA_STATE)) {
            state = (State<?>) getIntent().getSerializableExtra(EXTRA_STATE);
        } else {
            state = createState();
        }
        return state;
    }

    @SuppressLint("VisibleForTests")
    protected void applyState(@Nullable State<?> state) {
        final Navigable navigable = navigateTo();
        if (navigable == null) {
            throw new IllegalStateException(Navigable.class.getSimpleName() + " is null");
        }
        if (state != null) {
            navigable.state(Add.deeper(state));
        }
    }
}

File: jugglerhelper/src/main/kotlin/net/maxsmr/jugglerhelper/activities/BaseJugglerActivity.kt
package net.maxsmr.jugglerhelper.activities

import android.content.res.AssetManager
import android.os.Bundle
import android.view.KeyEvent
import android.view.MotionEvent
import androidx.annotation.AnimRes
import androidx.fragment.app.Fragment
import me.ilich.juggler.Juggler
import me.ilich.juggler.gui.JugglerActivity
import me.ilich.juggler.states.State
import net.maxsmr.jugglerhelper.fragments.BaseJugglerFragment
import net.maxsmr.jugglerhelper.utils.FragmentSearchParams
import net.maxsmr.jugglerhelper.utils.findFragment

abstract class BaseJugglerActivity : JugglerActivity() {

    @AnimRes
    protected open val finishEnterAnimation: Int = 0

    @AnimRes
    protected open val finishExitAnimation: Int = 0

    protected val activityState: State<*>?
        get() {
            val intent = intent
                    ?: throw IllegalStateException("Intent is null")
            val state: State<*>
            state = if (intent.hasExtra(EXTRA_STATE)) {
                getIntent().getSerializableExtra(EXTRA_STATE) as State<*>
            } else {
                createState()
            }
            return state
        }

    var isCommitAllowed = false
        private set

    var isActivityResumed = false
        private set

    protected var initialState: State<*>? = null
        private set

    protected var savedInstanceState: Bundle? = null
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        intent.putExtra(Juggler.DATA_ANIMATION_FINISH_ENTER, finishEnterAnimation)
        intent.putExtra(Juggler.DATA_ANIMATION_FINISH_EXIT, finishExitAnimation)
        initialState = activityState
        intent.putExtra(EXTRA_STATE, overrideInitialState(initialState))
        this.savedInstanceState = savedInstanceState
        isCommitAllowed = true
        super.onCreate(savedInstanceState)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        savedInstanceState = outState
        isCommitAllowed = false
    }

    override fun onResumeFragments() {
        super.onResumeFragments()
        isCommitAllowed = true
    }

    override fun onPause() {
        super.onPause()
        isActivityResumed = false
    }

    override fun onResume() {
        super.onResume()
        isActivityResumed = true
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        for (f in supportFragmentManager.fragments) {
            if (f is BaseJugglerFragment && f.isAdded()) {
                f.onTouchEvent(event)
            }
        }
        return super.onTouchEvent(event)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        for (f in supportFragmentManager.fragments) {
            if (f is BaseJugglerFragment && f.isAdded()) {
                f.onKeyDown(keyCode, event)
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        for (f in supportFragmentManager.fragments) {
            if (f != null && f.isAdded) {
                f.onRequestPermissionsResult(requestCode, permissions, grantResults)
            }
        }
    }

    override fun getAssets(): AssetManager {
        // при переопределении конфигурации (например, в LocaleContextWrapper или applyOverrideConfiguration)
        // возвращает другой инстанс в Context.getAssets() и Context.getResources().getAssets()
        // и не находит там нужные строки
        // начиная с версии appCompat 1.3
        return getResources().getAssets()
    }

    fun <F : Fragment> findFragment(params: FragmentSearchParams<F>)= findFragment(supportFragmentManager, params)

    protected open fun overrideInitialState(state: State<*>?): State<*>? = state
}
